src/hooks/useConnectionStatus.tsx

```diff
@@ -0,0 +1,97 @@
+import React, { useContext, useEffect, useState } from 'react'
+
+export type ConnectionStatus = 'connected' | 'disconnected' | 'connecting'
+
+const ConnectionStatusContext = React.createContext<undefined | ConnectionStatusValue>(undefined)
+ConnectionStatusContext.displayName = 'ConnectionStatusContext'
+
+type ConnectionStatusValue = {
+  setNodeConnection: (connection: ConnectionStatus) => void
+  nodeConnectionStatus: ConnectionStatus
+  isUserConnectedToInternet: boolean
+}
+
+export const ConnectionStatusProvider: React.FC = ({ children }) => {
+  const [nodeConnectionStatus, setNodeConnection] = useState<ConnectionStatus>('connecting')
+  const [navigatorOnline, setNavigatorOnline] = useState(true)
+  const [isUserConnectedToInternet, setIsUserConnectedToInternet] = useState(true)
+
+  useEffect(() => {
+    let timeout: NodeJS.Timeout
+    let interval: NodeJS.Timeout
+    const connectionHandler = () => {
+      setNavigatorOnline(window.navigator.onLine)
+    }
+
+    window.addEventListener('offline', connectionHandler)
+    window.addEventListener('online', connectionHandler)
+
+    return () => {
+      clearInterval(interval)
+      clearTimeout(timeout)
+      window.removeEventListener('offline', connectionHandler)
+      window.removeEventListener('online', connectionHandler)
+    }
+  }, [])
+
+  useEffect(() => {
+    if (!navigatorOnline) {
+      return
+    }
+    // when navigator.online ping google every three seconds to check if user is connected to internet
+    const interval = setInterval(() => {
+      checkConnection()
+    }, 3000)
+    return () => {
+      clearInterval(interval)
+    }
+  }, [navigatorOnline])
+
+  useEffect(() => {
+    if (navigatorOnline) {
+      return
+    }
+    const timeout = setTimeout(() => {
+      setIsUserConnectedToInternet(false)
+    }, 3000)
+    return () => {
+      clearInterval(timeout)
+    }
+  }, [navigatorOnline])
```

Comment:
Could you explain this approach? Do we need any additional logic apart from adding `setIsUserConnectedToInternet(false)` at line 39?